polyfill/intersectionobserver-polyfill.js

```diff
@@ -0,0 +1,188 @@
+(function() {
+  // TODO: rootMargin are completely ignored for now
+  
+  // Constructor
+  window.IntersectionObserver = function(callback, options) {
+    options = options || {};
+    
+    if(!(callback instanceof Function)) {
+      throw('callback needs to be a function');
+    }
+    
+    this._callback = callback;
+    this._root = options.root || null;
+    this._rootMargin = options.rootMargin || [0, 0, 0, 0];
+    this._thresholds = options.threshold || 0;
+    this._init();
+  };
+
+  window.IntersectionObserver.prototype = {
+    //
+    // Public API
+    //
+    get root() {
+      return this._root || document;
+    },
+    
+    get rootMargin() {
+      return '0';
+    },
+    
+    get thresholds() {
+      // 0 means call callback on every change
+      // See note at http://rawgit.com/WICG/IntersectionObserver/master/index.html#intersection-observer-init
+      if(this._thresholds === 0) {
+        return 0;
+      }
+      if(this._thresholds instanceof Array) {
+        return this._thresholds;
+      }
+      return [this._thresholds];
+    },
+    
+    observe: function(target) {
+      if(!(target instanceof HTMLElement)) {
+        throw('Target needs to be an HTMLelement');
+      }
+      
+      var root = this.root;
+      var ancestor = target.parentNode;
+      while (ancestor != root) {
+        if (!ancestor) {
+          throw('Target must be decendant of root');
+        }
+        ancestor = ancestor.parentNode;   
+      }
+      
+      this._observationTargets.set(target, {});
+    },
+    
+    unobserve: function(target) {
+      this._observationTargets.delete(target);
+    },
+    
+    disconnect: function() {
+      this._observationTargets.clear();
+      this.root.removeEventListener('scroll', this._boundUpdate);
+      window.clearInterval(this._timeoutId);
+      this._descheduleCallback();
+    },
+    
+    takeRecords: function() {
+      this._update();
+      this._descheduleCallback();
+      var copy = this._queue.slice();
+      this._queue = [];
+      return copy;
+    },
+    
+    //
+    // Private API
+    //
+    _init: function() {
+      this._observationTargets = new Map();
+      this._boundUpdate = this._update.bind(this);
+      this.root.addEventListener('scroll', this._boundUpdate);
+      this._intervalId = window.setInterval(this._boundUpdate, 100);
```

Comment:
This seems less efficient than typical lazyload libraries, in that it polls the state of the observed element on every scroll event, and every 100ms. I would like to put this polyfill into the polyfill.io service library but am concerned about the frequency of these events. How about throttling the callback to one invocation per 100ms?

Also I assume we're polling separately to the scroll events to capture scenarios where the element moves from outside to inside the viewport without the document scroll position changing. Could this lean on MutationObserver to avoid setting up a tight interval timer that runs indefinitely?